test/sqllogictest/transform/table_elimination.slt

```diff
@@ -0,0 +1,19 @@
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
```

Comment:
It seems that the optimization kicks in only for self-joins. If the second table is just like the first, but actually a different table, the optimization does not kick in:

```
CREATE TABLE t1 (key integer PRIMARY KEY, nokey integer);
CREATE TABLE t2 (key integer PRIMARY KEY, nokey integer);

EXPLAIN SELECT a1.* FROM t1 AS a1 LEFT JOIN t1 AS a2 ON (a1.key = a2.key);
```

This produces a full blown plan:

```
materialize=> EXPLAIN SELECT a1.* FROM t1 AS a1 LEFT JOIN t2 AS a2 ON (a1.key = a2.key);
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.t1 (u71)+
 | ArrangeBy (#0)                 +
                                  +
 %1 =                             +
 | Get materialize.public.t2 (u73)+
 | ArrangeBy (#0)                 +
                                  +
 %2 = Let l0 =                    +
 | Join %0 %1 (= #0 #2)           +
 | | implementation = DeltaQuery  +
 | |   delta %0 %1.(#0)           +
 | |   delta %1 %0.(#0)           +
 | | demand = (#0, #1)            +
                                  +
 %3 =                             +
 | Get %2 (l0)                    +
 | Negate                         +
 | Project (#0, #1)               +
                                  +
 %4 =                             +
 | Get materialize.public.t1 (u71)+
                                  +
 %5 =                             +
 | Union %3 %4                    +
 | Map null, null                 +
                                  +
 %6 =                             +
 | Union %5 %2                    +
 | Project (#0, #1)               +
```

Reply:
I don't think that case is supposed to be covered by the proposed transformation.

Wouldn't this be better done at the semantic level?

Reply:
I agree this is out of scope. I have created #6449 to track this issue.